nixos/modules/system/activation/bootspec.nix

```diff
@@ -0,0 +1,56 @@
+# Note that these schemas are defined by RFC-0125.
+# This document is considered a stable API, and is depended upon by external tooling.
+# Changes to the structure of the document, or the semantics of the values should go through an RFC.
+#
+# See: https://github.com/NixOS/rfcs/pull/125
+{ config, pkgs, lib, children }:
+let
+  schemas = {
+    v1 = rec {
+      filename = "boot.v1.json";
+      json =
+        pkgs.writeText filename
+          (builtins.toJSON
+            {
+              schemaVersion = 1;
+
+              kernel = "${config.boot.kernelPackages.kernel}/${config.system.boot.loader.kernelFile}";
+              kernelParams = config.boot.kernelParams;
+              initrd = "${config.system.build.initialRamdisk}/${config.system.boot.loader.initrdFile}";
+              initrdSecrets = "${config.system.build.initialRamdiskSecretAppender}/bin/append-initrd-secrets";
+              label = "NixOS ${config.system.nixos.codeName} ${config.system.nixos.label} (Linux ${config.boot.kernelPackages.kernel.modDirVersion})";
+            });
+
+      generator =
+        let
+          specialisationLoader = (lib.mapAttrsToList
+            (childName: childToplevel: lib.escapeShellArgs [ "--slurpfile" childName "${childToplevel}/${filename}" ])
+            children);
+        in
+        ''
+          mkdir -p $out/bootspec
+
+          ${pkgs.jq}/bin/jq '
+            .toplevel = $toplevel |
+            .init = $init
+            ' \
+            --sort-keys \
+            --arg toplevel "$out" \
+            --arg init "$out/init" \
```

Comment:
I wouldn't mind a comment here what is being sorted. On first look we substitute top-level and init and then select specializations from the json.

Reply:
@cole-h We still need to address this with a comment, I think it can be moved somewhere safe.

Reply:
Open to suggestions :)

Reply:
https://github.com/NixOS/nixpkgs/pull/172237/files#diff-097959078eaf4721c05782835844d922b4810ac16f8aec1598881036648ba36aR43

with the comment you dropped in the last commit, I think, would be a good place!

Reply:
I have a better idea, coming soon (if it works). (What I don't  like about just moving it up there is that it's far from where the specialisation stuff is happening, so it's difficult to immediately know what it's talking about.)

Reply:
[`bbada7c` (#172237)](https://github.com/NixOS/nixpkgs/pull/172237/commits/bbada7cb097e2986e2233b295f2dacae82b44350)

Reply:
Great!